File: lib/src/main/java/com/sd/lib/http/cookie/SerializableCookieStore.java
package com.sd.lib.http.cookie;

import android.content.Context;

import com.sd.lib.http.utils.HttpIOUtil;
import com.sd.lib.http.utils.HttpLog;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

/**
 * 通过序列化和反序列化实现的可持久化CookieStore
 */
public class SerializableCookieStore extends PersistentCookieStore
{
    private Context mContext;
    private ModifyMemoryCookieStore mMemoryCookieStore;
    private File mFile;

    public SerializableCookieStore(Context context)
    {
        mContext = context.getApplicationContext();
    }

    @Override
    protected synchronized ModifyMemoryCookieStore getMemoryCookieStore()
    {
        if (mMemoryCookieStore == null)
        {
            try
            {
                mMemoryCookieStore = deserializeObject(getFile());
            } catch (Exception e)
            {
                HttpLog.e("cookie deserialize cookiestore error:" + e);
            }

            if (mMemoryCookieStore == null)
            {
                mMemoryCookieStore = new ModifyMemoryCookieStore();
                HttpLog.i("cookie create MemoryCookieStore");
            } else
            {
                HttpLog.i("cookie deserialize cookiestore success");
            }
        }
        return mMemoryCookieStore;
    }

    private File getFile()
    {
        if (mFile == null)
        {
            mFile = new File(mContext.getFilesDir(), "httpcookie");
        }
        if (!mFile.exists())
        {
            try
            {
                mFile.createNewFile();
            } catch (Exception e)
            {
                HttpLog.e("cookie create httpcookie file error:" + e);
            }
        }
        return mFile;
    }

    @Override
    protected synchronized void save()
    {
        try
        {
            serializeObject(getMemoryCookieStore(), getFile());
            HttpLog.i("cookie save cookiestore success");
        } catch (Exception e)
        {
            HttpLog.e("cookie save cookiestore error:" + e);
        }
    }

    private static <T extends Serializable> void serializeObject(T object, File file) throws Exception
    {
        ObjectOutputStream oos = null;
        try
        {
            oos = new ObjectOutputStream(new FileOutputStream(file));
            oos.writeObject(object);
            oos.flush();
        } finally
        {
            HttpIOUtil.closeQuietly(oos);
        }
    }

    private static <T extends Serializable> T deserializeObject(File file) throws Exception
    {
        ObjectInputStream ois = null;
        try
        {
            ois = new ObjectInputStream(new FileInputStream(file));
            Object object = ois.readObject();
            return (T) object;
        } finally
        {
            HttpIOUtil.closeQuietly(ois);
        }
    }
}

File: lib/src/main/java/com/sd/lib/http/cookie/store/SerializableCookieStore.kt
package com.sd.lib.http.cookie.store

import android.content.Context
import com.sd.lib.http.cookie.SerializableHttpCookie
import com.sd.lib.http.utils.HttpIOUtils
import com.sd.lib.http.utils.HttpLog
import java.io.*
import java.net.URI

class SerializableCookieStore : PersistentCookieStore {

    private var _file: File? = null
        get() {
            if (field == null) {
                field = File(context.filesDir, "f_http_cookie")
            }
            return field
        }

    constructor(context: Context) : super(context)

    @Synchronized
    override fun saveCacheImpl(cookies: Map<URI?, List<SerializableHttpCookie>>) {
        try {
            val file = _file!!
            if (!file.exists()) {
                file.createNewFile()
            }

            val holder = SerializableCookieHolder().apply {
                this.map = cookies
            }

            serializeObject(holder, file)
            HttpLog.i("cookie CookieStore saveCacheImpl success")
        } catch (e: Exception) {
            HttpLog.e("cookie CookieStore saveCacheImpl error:${e}")
        }
    }

    override fun getCacheImpl(): Map<URI?, List<SerializableHttpCookie>>? {
        try {
            val file = _file!!
            val obj = deserializeObject<SerializableCookieHolder>(file)
            HttpLog.i("cookie CookieStore getCacheImpl obj:${obj}")
            return obj?.map
        } catch (e: Exception) {
            HttpLog.e("cookie CookieStore getCacheImpl error:${e}")
            return null
        }
    }

    companion object {
        private fun <T : Serializable> serializeObject(obj: T, file: File) {
            var oos: ObjectOutputStream? = null
            try {
                oos = ObjectOutputStream(FileOutputStream(file))
                oos.writeObject(obj)
                oos.flush()
            } finally {
                HttpIOUtils.closeQuietly(oos)
            }
        }

        private fun <T : Serializable> deserializeObject(file: File): T {
            var ois: ObjectInputStream? = null
            try {
                ois = ObjectInputStream(FileInputStream(file))
                return ois.readObject() as T
            } finally {
                HttpIOUtils.closeQuietly(ois)
            }
        }
    }
}

class SerializableCookieHolder : Serializable {
    var map: Map<URI?, List<SerializableHttpCookie>>? = null

    companion object {
        private const val serialVersionUID = 42L
    }
}